addons/hr_payroll/models/hr_benefit.py

```diff
@@ -0,0 +1,271 @@
+                days.append((current_start, current_end))
+                current_start = current_end + relativedelta(seconds=1)
+                current_end = current_end + relativedelta(days=1)
+
+            days.append((current_start, end))
+
+            # filter to avoid dummy intervals starting and ending at the same time
+            return [(start, end) for start, end in days if start != end]
+
+        self.ensure_one()
+        if self.date_start.date() == self.date_stop.date():
+            return self
+
+        tz = pytz.timezone(self.employee_id.tz)
+        benefit_start, benefit_stop = tz.localize(self.date_start), tz.localize(self.date_stop)
+        values = {
+            'name': self.name,
+            'employee_id': self.employee_id.id,
+            'benefit_type_id': self.benefit_type_id.id,
+            'state': self.state
+        }
+        self.unlink()
+        new_benefits = self.env['hr.benefit']
+        for start, stop in _split_range_by_day(benefit_start, benefit_stop):
+            values['date_start'] = start.astimezone(pytz.utc)
+            values['date_stop'] = stop.astimezone(pytz.utc)
+            new_benefits |= self.create(values)
+        return new_benefits
+
+    @api.multi
+    def _duplicate_to_calendar(self):
+        """
+            Duplicate data to keep the complexity in benefit and not mess up payroll, etc.
+        """
+        attendance_benefits = self.filtered(lambda b:
+            not b.benefit_type_id.is_leave and
+            # Normal benefit are global to all employees -> avoid duplicating it
+            not b.benefit_type_id == self.env.ref('hr_payroll.benefit_type_attendance'))
+        leave_benefits = self.filtered(lambda b: b.benefit_type_id.is_leave)
+
+        for benefit in attendance_benefits:
+            benefit = benefit.split_by_day()
+            benefit._duplicate_to_calendar_attendance()
+        leave_benefits._duplicate_to_calendar_leave()
+
+    @api.multi
+    def _duplicate_to_calendar_leave(self):
+
+        for benefit in self:
+            if not benefit.leave_id:
+                tz = pytz.timezone(benefit.employee_id.tz)
+                self.env['resource.calendar.leaves'].create({
+                    'name': benefit.name,
+                    'date_from': benefit.date_start,
+                    'date_to': benefit.date_stop,
+                    'calendar_id': benefit.employee_id.resource_calendar_id.id,
+                    'resource_id': benefit.employee_id.resource_id.id,
+                    'benefit_type_id': benefit.benefit_type_id.id,
+                })
+
+    @api.multi
+    def _duplicate_to_calendar_attendance(self):
+
+        for benefit in self:
+            tz = pytz.timezone(benefit.employee_id.tz)
+            start = pytz.utc.localize(benefit.date_start).astimezone(tz) # in UTC -> timezoned
+            end = pytz.utc.localize(benefit.date_stop).astimezone(tz)
+
+            assert start.date() == end.date()
```

Comment:
Careful, this assertion will break the loop and rollback what has already been done
Maybe filter out these beforehand